EventSource/src/EventSourceUtil.ts

```diff
@@ -57,6 +57,129 @@ export function getTimeStamp(resource: Resource, timestampPath: string): number
+    const blank = new Map<string, Map<string, Quad_Object[]>>();
+    addElements:
+    for (const quad of resource) {
+        const data = quad.subject.termType == "BlankNode" ? blank : named;
+        if (data.has(quad.subject.id)) {
+            const props = data.get(quad.subject.id)!;
+            if (props.has(quad.predicate.id)) {
+                // check if value is already in array, if it is, dont add it anymore
+                const objs = props.get(quad.predicate.id)!;
+                for (const obj of objs) {
+                    // while it might offer better performance to use a set instead
+                    // of an array, the custom type Quad_Object would not work correctly
+                    // with Set.has(), and thus would require a seperate container storing
+                    // the IDs (which would in turn not be memory efficient)
+                    if (obj.equals(quad.object)) {
+                        continue addElements;
+                    }
+                }
+                objs.push(quad.object);
+            } else {
+                props.set(quad.predicate.id, new Array(quad.object));
+            }
+        } else {
+            data.set(quad.subject.id, new Map([[quad.predicate.id, new Array(quad.object)]]));
+        }
+    }
+    return [named, blank];
+}
+
+/**
+ * Converts a resource (quad array) to an optimised turtle string representation by grouping subjects
+ * together, using prefixes wherever possible and replacing blank nodes with their properties.
+ * Note: blank nodes referenced as objects, but not found as subjects in other quads, are removed
+ *  entirely
+ *
+ * @param resource The resource that gets converted to a string
+ * @param _prefixes An object which members are strings, member name being the short prefix and its
+ *  value a string representing its URI. Example: `{"rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#"}`
+ * @returns {string}
+ */
+export function resourceToOptimisedTurtle(resource: Resource, _prefixes: any): string {
+    // get a grouped overview of this resource's content
+    const [named, blank] = resourceToMaps(resource);
+    // converting all the entries of the blank map first
+    const blankEntries = new Map<string, {predicate: NamedNode, objects: Quad_Object[]}[]>();
+    for (const [subject, properties] of blank) {
+        blankEntries.set(subject, []);
+        for (const [property, objects] of properties) {
+            blankEntries.get(subject)!.push({
+                predicate: namedNode(property),
+                objects: objects
+            });
+        }
+    }
+    // with the ordered view done, a more compact turtle string can be generated
+    const writer = new Writer({prefixes: _prefixes});
+    // adding all the blank nodes with their properties first
+    for (const [subject, properties] of named) {
+        for (const [predicate, objects] of properties) {
+            for (const object of objects) {
```

Comment:
Is there a big difference in output when going over every triple like this and just using Writer with prefixes without going through the loops?

Like, does `writer.addQuads(resource)` produce something else than what the whole function does?

Reply:
I based this logic on the example found at https://github.com/rdfjs/N3.js#blank-nodes-and-lists, where the description reads that "a streaming writer cannot create [short representations of blank nodes] automatically: the shorthand notations are only possible if blank nodes or list heads are not used later on, which can only be determined conclusively at the end of the stream", and interpreted this such that this flow would be required. I haven't tested manually just adding resources, as that is what the original implementation does I believe (storeToString) and didn't yield the desired result. I haven't tested this however, so I will check it and report my findings back tomorrow.

Reply:
The following code produces a slightly different (and less optimized) layout:
`writer.addQuads(resource);
 let result;
 writer.end((err, res) => result = res);
 return result;`
The notable difference in at least every possible layout for a grouped resource is the declaration of every group member in the event stream:
`:EventStream tree:member <https://example.com/sample1>, <https://example.com/sample2>, [...] .`
becomes
`:EventStream tree:member <https://example.com/sample1>.
[...]
:EventStream tree:member <https://example.com/sample2>.
...`
This comes with a file size increase of around 32% (5.4MB instead of 4.1) in my use case, which is not desirable.